python/private/pypi/uv_extension.bzl

```diff
@@ -0,0 +1,60 @@
+# Copyright 2024 The Bazel Authors. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"uv module extension"
+
+load("//python:repositories.bzl", "uv_register_toolchains")
+
+_DEFAULT_NAME = "uv"
+
+uv_toolchain = tag_class(attrs = {
+    "name": attr.string(doc = """\
+Base name for generated repositories, allowing more than one uv toolchain to be registered.
+Overriding the default is only permitted in the root module.
+""", default = "uv"),
+    "uv_version": attr.string(doc = "Explicit version of uv.", mandatory = True),
+})
+
+def _uv_toolchain_extension(module_ctx):
+    registrations = {}
+    for mod in module_ctx.modules:
+        for toolchain in mod.tags.toolchain:
+            if toolchain.name != _DEFAULT_NAME and not mod.is_root:
+                fail("""\
+                Only the root module may override the default name for the uv toolchain.
+                This prevents conflicting registrations in the global namespace of external repos.
+                """)
+            if toolchain.name not in registrations.keys():
+                registrations[toolchain.name] = []
+            registrations[toolchain.name].append(toolchain.uv_version)
+    for name, versions in registrations.items():
+        if len(versions) > 1:
+            # TODO: should be semver-aware, using MVS
```

Comment:
I don't have a strong opinion about whether, internally, some MVS thing is done over all the requested toolchains. Just note it has incomplete information -- a user can register a toolchain outside of calling `uv.toolchain`, which will have higher priority, and won't be seen by the uv extension.